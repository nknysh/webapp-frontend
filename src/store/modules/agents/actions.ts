import { ICompany, ITravelAgent } from 'services/BackendApi';
export const GET_TRAVEL_AGENTS_REQUEST = 'agentsModule/GET_TRAVEL_AGENTS_REQUEST';
export const GET_TRAVEL_AGENTS_SUCCESS = 'agentsModule/GET_TRAVEL_AGENTS_SUCCESS';
export const GET_TRAVEL_AGENTS_FAILURE = 'agentsModule/GET_TRAVEL_AGENTS_FAILURE';
export const SELECTED_TA_CHANGE = 'agentsModule/SELECTED_TA_CHANGE';
export const SHOW_TA_DROPDOWN = 'agentsModule/SHOW_TA_DROPDOWN';
export const SEARCH_TA_BY_NAME_CHANGE = 'agentsModule/SEARCH_TA_BY_NAME_CHANGE';
export const GET_COMPANIES_REQUEST = 'agentsModule/GET_COMPANIES_REQUEST';
export const GET_COMPANIES_SUCCESS = 'agentsModule/GET_COMPANIES_SUCCESS';
export const GET_COMPANIES_FAILURE = 'agentsModule/GET_COMPANIES_FAILURE';
export const SELECTED_COMPANY_CHANGE = 'agentsModule/SELECTED_COMPANY_CHANGE';
export const SHOW_COMPANY_DROPDOWN = 'agentsModule/SHOW_COMPANY_DROPDOWN';
export const SEARCH_COMPANY_BY_NAME = 'agentsModule/SEARCH_COMPANY_BY_NAME';

export type GetTravelAgentRequestAction = ReturnType<typeof getTravelAgentsRequestAction>;
export const getTravelAgentsRequestAction = (forceLoad?: boolean) => ({
  type: GET_TRAVEL_AGENTS_REQUEST as typeof GET_TRAVEL_AGENTS_REQUEST,
  forceLoad,
});

export type GetTravelAgentSuccessAction = ReturnType<typeof getTravelAgentsSuccessAction>;
export const getTravelAgentsSuccessAction = (agents: ITravelAgent[]) => ({
  type: GET_TRAVEL_AGENTS_SUCCESS as typeof GET_TRAVEL_AGENTS_SUCCESS,
  agents,
});

export type GetTravelAgentFailureAction = ReturnType<typeof getTravelAgentsFailureAction>;
export const getTravelAgentsFailureAction = (error: string) => ({
  type: GET_TRAVEL_AGENTS_FAILURE as typeof GET_TRAVEL_AGENTS_FAILURE,
  error,
});

export type SelectedTaChangeAction = ReturnType<typeof selectedTaChangeAction>;
export const selectedTaChangeAction = (value: ITravelAgent | null) => ({
  type: SELECTED_TA_CHANGE as typeof SELECTED_TA_CHANGE,
  value,
});

export type ShowTaDropdownAction = ReturnType<typeof showTaDropdownAction>;
export const showTaDropdownAction = (value: boolean) => ({
  type: SHOW_TA_DROPDOWN as typeof SHOW_TA_DROPDOWN,
  value,
});

export type SearchTaByNameChangeAction = ReturnType<typeof searchTaByNameChangeAction>;
export const searchTaByNameChangeAction = (value: string) => ({
  type: SEARCH_TA_BY_NAME_CHANGE as typeof SEARCH_TA_BY_NAME_CHANGE,
  value,
});


export type GetCompaniesRequestAction = ReturnType<typeof getCompaniesRequestAction>;
export const getCompaniesRequestAction = () => ({
  type: GET_COMPANIES_REQUEST as typeof GET_COMPANIES_REQUEST,
});

export type GetCompaniesSuccessAction = ReturnType<typeof getCompaniesSuccessAction>;
export const getCompaniesSuccessAction = (companies: ICompany[]) => ({
  type: GET_COMPANIES_SUCCESS as typeof GET_COMPANIES_SUCCESS,
  companies,
});

export type GetCompaniesFailureAction = ReturnType<typeof getCompaniesFailureAction>;
export const getCompaniesFailureAction = (error: string) => ({
  type: GET_COMPANIES_FAILURE as typeof GET_COMPANIES_FAILURE,
  error,
});

export type SelectedCompanyChangeAction = ReturnType<typeof selectedCompanyChangeAction>;
export const selectedCompanyChangeAction = (value: ICompany) => ({
  type: SELECTED_COMPANY_CHANGE as typeof SELECTED_COMPANY_CHANGE,
  value,
});

export type ShowCompanyDropdownAction = ReturnType<typeof showCompanyDropdownAction>;
export const showCompanyDropdownAction = (value: boolean) => ({
  type: SHOW_COMPANY_DROPDOWN as typeof SHOW_COMPANY_DROPDOWN,
  value,
});

export type SearchCompanyByNameAction = ReturnType<typeof searchCompanyByNameAction>;
export const searchCompanyByNameAction = (value: string) => ({
  type: SEARCH_COMPANY_BY_NAME as typeof SEARCH_COMPANY_BY_NAME,
  value,
});

export type AgentsModuleAction =
  | GetTravelAgentRequestAction
  | GetTravelAgentSuccessAction
  | GetTravelAgentFailureAction
  | SelectedTaChangeAction
  | ShowTaDropdownAction
  | SearchTaByNameChangeAction
  | GetCompaniesRequestAction
  | GetCompaniesSuccessAction
  | GetCompaniesFailureAction
  | SelectedCompanyChangeAction
  | ShowCompanyDropdownAction
  | SearchCompanyByNameAction;
