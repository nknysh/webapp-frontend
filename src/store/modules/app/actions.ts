export const SET_LATEST_APP_VERSION = 'app/SET_LATEST_APP_VERSION';

export type SetLatestAppVersion = ReturnType<typeof setLatestAppVersion>;
export const setLatestAppVersion = (latestAppVersion: string, isAppDeprecated: boolean) => ({
  type: SET_LATEST_APP_VERSION as typeof SET_LATEST_APP_VERSION,
  latestAppVersion,
  isAppDeprecated
});


export type AppAction =
 | SetLatestAppVersion
