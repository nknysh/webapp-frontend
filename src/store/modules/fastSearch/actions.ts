import { Filters, OffersSearchSuccessResponse, ErrorResponse } from 'services/BackendApi';
import { SearchQuery, SearchOptions, StarRating } from 'services/BackendApi/types';
import { Occasion } from 'services/BackendApi/types/SearchQuery';
import { MealPlan } from '../../../services/BackendApi/types/OffersSearchResponse';
import { MealPlanNames } from '../../../services/BackendApi/types/SearchQuery';

export const DESTINATION_CHANGE = 'fastSearch/DESTINATION_CHANGE';
export const TOGGLE_FILTER = 'fastSearch/TOGGLE_FILTER';
export const SET_FILTERS = 'fastSearch/SET_FILTER';
export const SET_ALL_FILTERS = 'fastSearch/SET_ALL_FILTERS';
export const TOGGLE_REPEAT_GUEST = 'fastSearch/TOGGLE_REPEAT_GUEST';
export const TOGGLE_STAR_RATING = 'fastSearch/TOGGLE_STAR_RATING';
export const TOGGLE_OCCASION = 'fastSearch/TOGGLE_OCCASION';
export const TOGGLE_REGION = 'fastSearch/TOGGLE_REGION';
export const TOGGLE_SHOW_REGIONS = 'fastSearch/TOGGLE_SHOW_REGIONS';
export const SELECT_MEAN_PLAN = 'fastSearch/SELECT_MEAL_PLAN';
export const DATE_RANGE_START_CHANGE = 'fastSearch/DATE_RANGE_START_CHANGE';
export const DATE_RANGE_END_CHANGE = 'fastSearch/DATE_RANGE_END_CHANGE';
export const MIN_PRICE_CHANGE = 'fastSearch/MIN_PRICE_CHANGE';
export const MAX_PRICE_CHANGE = 'fastSearch/MAX_PRICE_CHANGE';
export const INCREMENT_ROOM = 'fastSearch/INCREMENT_ROOM';
export const DECREMENT_ROOM = 'fastSearch/DECREMENT_ROOM';
export const INCREMENT_ADULT = 'fastSearch/INCREMENT_ADULT';
export const DECREMENT_ADULT = 'fastSearch/DECREMENT_ADULT';
export const INCREMENT_CHILD = 'fastSearch/INCREMENT_CHILD';
export const DECREMENT_CHILD = 'fastSearch/DECREMENT_CHILD';
export const AGE_CHANGE = 'fastSearch/AGE_CHANGE';
export const OFFERS_SEARCH_REQUEST = 'fastSearch/OFFERS_SEARCH_REQUEST';
export const OFFERS_SEARCH_SUCCESS = 'fastSearch/OFFERS_SEARCH_SUCCESS';
export const OFFERS_SEARCH_FAILURE = 'fastSearch/OFFERS_SEARCH_FAILURE';
export const OPTIONS_REQUEST = 'fastSearch/OPTIONS_REQUEST';
export const OPTIONS_SUCCESS = 'fastSearch/OPTIONS_SUCCESS';
export const OPTIONS_FAILURE = 'fastSearch/OPTIONS_FAILURE';

// ---------------------------------------------------------------------------------------
// How to type Redux Actions with as little boilerplate as possible.
// ---------------------------------------------------------------------------------------
//
// The objective here is to have the types driven by the implementation, so you can
// edit and change actions easily, and get all the benefits of Typescript automatically
// without having to manually create and maintain types for everything. ProTip: Make
// sure your actions are always simple object literals. That means, no "thunks". This
// method is a lot more boilerplatey than before, but it's a lot easier to figure out
// what's happening and to make changes.
//
// 1. Infer the ReturnType of the action. We'll use this to create a Union type later.
export type DestinationChangeAction = ReturnType<typeof destinationChangeAction>;
export const destinationChangeAction = (value: string) => ({
  // 2. Infer the string literal type of the action. This helps typescript narrow down the
  //    Action type in the any reducers Switch statemet.
  //    See `FastSearchAction` at the bottom of this file for step 3.
  type: DESTINATION_CHANGE as typeof DESTINATION_CHANGE,
  value,
});
// ---------------------------------------------------------------------------------------

export type ToggleFilterAction = ReturnType<typeof toggleFilterAction>;
export const toggleFilterAction = (filter: Filters) => ({
  type: TOGGLE_FILTER as typeof TOGGLE_FILTER,
  filter,
});

export type SetFiltersAction = ReturnType<typeof setFiltersAction>;
export const setFiltersAction = (filters: Filters[], value: boolean) => ({
  type: SET_FILTERS as typeof SET_FILTERS,
  filters,
  value,
});

export type SetAllFiltersAction = ReturnType<typeof setAllFiltersAction>;
export const setAllFiltersAction = (value: boolean) => ({
  type: SET_ALL_FILTERS as typeof SET_ALL_FILTERS,
  value,
});

export type ToggleRepeatGuestAction = ReturnType<typeof toggleRepeatGuestAction>;
export const toggleRepeatGuestAction = () => ({
  type: TOGGLE_REPEAT_GUEST as typeof TOGGLE_REPEAT_GUEST,
});

export type ToggleStarRatingAction = ReturnType<typeof toggleStarRatingAction>;
export const toggleStarRatingAction = (starRating: StarRating) => ({
  type: TOGGLE_STAR_RATING as typeof TOGGLE_STAR_RATING,
  starRating,
});

export type ToggleOccasionAction = ReturnType<typeof toggleOccasionAction>;
export const toggleOccasionAction = (occasion: Occasion) => ({
  type: TOGGLE_OCCASION as typeof TOGGLE_OCCASION,
  occasion,
});

export type ToggleShowRegionsAction = ReturnType<typeof toggleShowRegionsAction>;
export const toggleShowRegionsAction = () => ({
  type: TOGGLE_SHOW_REGIONS as typeof TOGGLE_SHOW_REGIONS,
});

export type ToggleRegionAction = ReturnType<typeof toggleRegionAction>;
export const toggleRegionAction = (region: string) => ({
  type: TOGGLE_REGION as typeof TOGGLE_REGION,
  region,
});

export type SelectMealPlanAction = ReturnType<typeof selectMealPlanAction>;
export const selectMealPlanAction = (mealPlan: MealPlanNames) => ({
  type: SELECT_MEAN_PLAN as typeof SELECT_MEAN_PLAN,
  mealPlan,
});

export type DateRangeStartChangeAction = ReturnType<typeof dateRangeStartChangeAction>;
export const dateRangeStartChangeAction = (date: string) => ({
  type: DATE_RANGE_START_CHANGE as typeof DATE_RANGE_START_CHANGE,
  date,
});

export type DateRangeEndChangeAction = ReturnType<typeof dateRangeEndChangeAction>;
export const dateRangeEndChangeAction = (date: string) => ({
  type: DATE_RANGE_START_CHANGE as typeof DATE_RANGE_START_CHANGE,
  date,
});

export type MinPriceChangeAction = ReturnType<typeof minPriceChangeAction>;
export const minPriceChangeAction = (value: number) => ({
  type: MIN_PRICE_CHANGE as typeof MIN_PRICE_CHANGE,
  value,
});

export type MaxPriceChangeAction = ReturnType<typeof maxPriceChangeAction>;
export const maxPriceChangeAction = (value: number) => ({
  type: MAX_PRICE_CHANGE as typeof MAX_PRICE_CHANGE,
  value,
});

export type IncrementRoomAction = ReturnType<typeof incrementRoomAction>;
export const incrementRoomAction = (lodgingIndex: number) => ({
  type: INCREMENT_ROOM as typeof INCREMENT_ROOM,
  lodgingIndex,
});

export type DecrementRoomAction = ReturnType<typeof decrementRoomAction>;
export const decrementRoomAction = (lodgingIndex: number) => ({
  type: DECREMENT_ROOM as typeof DECREMENT_ROOM,
  lodgingIndex,
});

export type IncrementAdultAction = ReturnType<typeof incrementAdultAction>;
export const incrementAdultAction = (lodgingIndex: number) => ({
  type: INCREMENT_ADULT as typeof INCREMENT_ADULT,
  lodgingIndex,
});

export type DecrementAdultAction = ReturnType<typeof decrementAdultAction>;
export const decrementAdultAction = (lodgingIndex: number) => ({
  type: DECREMENT_ADULT as typeof DECREMENT_ADULT,
  lodgingIndex,
});

export type IncrementChildAction = ReturnType<typeof incrementChildAction>;
export const incrementChildAction = (lodgingIndex: number) => ({
  type: INCREMENT_CHILD as typeof INCREMENT_CHILD,
  lodgingIndex,
});

export type DecrementChildAction = ReturnType<typeof decrementChildAction>;
export const decrementChildAction = (lodgingIndex: number) => ({
  type: DECREMENT_CHILD as typeof DECREMENT_CHILD,
  lodgingIndex,
});

export type AgeChangeAction = ReturnType<typeof ageChangeAction>;
export const ageChangeAction = (lodgingIndex: number, childIndex: number, value: string) => ({
  type: AGE_CHANGE as typeof AGE_CHANGE,
  lodgingIndex,
  childIndex,
  value,
});

export type SearchRequestAction = ReturnType<typeof offersSearchRequestAction>;
export const offersSearchRequestAction = (query: SearchQuery) => ({
  type: OFFERS_SEARCH_REQUEST as typeof OFFERS_SEARCH_REQUEST,
  query,
});

export type OffersSearchSuccessAction = ReturnType<typeof offersSearchSuccessAction>;
export const offersSearchSuccessAction = (successResponse: OffersSearchSuccessResponse) => ({
  type: OFFERS_SEARCH_SUCCESS as typeof OFFERS_SEARCH_SUCCESS,
  successResponse,
});

export type OffersSearchFailureAction = ReturnType<typeof offersSearchFailureAction>;
export const offersSearchFailureAction = (errorResponse: ErrorResponse) => ({
  type: OFFERS_SEARCH_FAILURE as typeof OFFERS_SEARCH_FAILURE,
  errorResponse,
});

export type OptionsRequestAction = ReturnType<typeof optionsRequestAction>;
export const optionsRequestAction = () => ({
  type: OPTIONS_REQUEST as typeof OPTIONS_REQUEST,
});

export type OptionsSuccessAction = ReturnType<typeof optionsSuccessAction>;
export const optionsSuccessAction = (successResponse: SearchOptions) => ({
  type: OPTIONS_SUCCESS as typeof OPTIONS_SUCCESS,
  successResponse,
});

export type OptionsFailureAction = ReturnType<typeof optionsFailureAction>;
export const optionsFailureAction = (errorResponse: ErrorResponse) => ({
  type: OPTIONS_FAILURE as typeof OPTIONS_FAILURE,
  errorResponse,
});

// 3. Create a union type which we can pass as the reducers action type.
// goto DestinationChangeAction to see start of these comments.
export type FastSearchAction =
  | DestinationChangeAction
  | ToggleFilterAction
  | ToggleRepeatGuestAction
  | ToggleStarRatingAction
  | ToggleOccasionAction
  | ToggleShowRegionsAction
  | ToggleRegionAction
  | SelectMealPlanAction
  | SetFiltersAction
  | SetAllFiltersAction
  | DateRangeStartChangeAction
  | DateRangeEndChangeAction
  | MinPriceChangeAction
  | MaxPriceChangeAction
  | IncrementRoomAction
  | DecrementRoomAction
  | IncrementAdultAction
  | DecrementAdultAction
  | IncrementChildAction
  | DecrementChildAction
  | AgeChangeAction
  | SearchRequestAction
  | OffersSearchSuccessAction
  | OffersSearchFailureAction
  | OptionsRequestAction
  | OptionsSuccessAction
  | OptionsFailureAction;
