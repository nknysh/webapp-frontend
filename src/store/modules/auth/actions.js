import { prop, propEq, omit, propOr, mergeDeepLeft, equals } from 'ramda';
import { FORBIDDEN } from 'http-status';

import client from 'api/auth';
import userClient from 'api/users';
import { UserStatusTypes, AuthTypes } from 'config/enums';

import { genericAction, successAction, errorAction, errorFromResponse, storeReset } from 'store/common';
import { enqueueNotification } from 'store/modules/ui/actions';
import { getCurrentUserType } from './selectors';

export const AUTH_REQUEST = 'AUTH_REQUEST';
export const AUTH_CHECK = 'AUTH_CHECK';
export const AUTH_OK = 'AUTH_OK';
export const AUTH_RESET = 'AUTH_RESET';
export const AUTH_SET_TOKEN = 'AUTH_SET_TOKEN';
export const AUTH_SIGN_UP = 'AUTH_SIGN_UP';
export const AUTH_LOG_OUT = 'AUTH_LOG_OUT';
export const AUTH_PASSWORD_RESET = 'AUTH_PASSWORD_RESET';
export const AUTH_SET_PASSWORD = 'AUTH_SET_PASSWORD';
export const AUTH_COUNTRY_SET = 'AUTH_COUNTRY_SET';
export const AUTH_PASSWORD_UPDATE = 'AUTH_PASSWORD_UPDATE';

// Localstorage constants for auth
export const AUTH_TOKEN = 'authToken';
export const AUTH_USER = 'authUser';
export const AUTH_COUNTRY_CODE = 'authCountryCode';

/**
 * Auth reset action
 *
 * @returns {object}
 */
const authReset = () => ({
  type: AUTH_RESET,
});

/**
 * Set token action
 *
 * @param {string} token
 * @returns {object}
 */
export const setToken = token => ({
  type: AUTH_SET_TOKEN,
  payload: { token },
});

/**
 * Auth request action
 *
 * @param {*} value
 * @returns {object}
 */
export const authRequest = value => ({
  type: AUTH_REQUEST,
  payload: value,
});

/**
 * Auth password reset action
 *
 * @param {*} value
 * @returns {object}
 */
export const authPasswordReset = value => ({
  type: AUTH_PASSWORD_RESET,
  payload: value,
});

/**
 * Auth set password reset action
 *
 * @param {*} value
 * @returns {object}
 */
export const authSetPasswordReset = value => ({
  type: AUTH_SET_PASSWORD,
  payload: value,
});

/**
 * Auth ok action
 *
 * @param {*} values
 * @returns {object}
 */
export const authOk = values => ({
  type: AUTH_OK,
  payload: values,
});

/**
 * Auth sign up action
 *
 * @param {*} values
 * @returns {object}
 */
export const authSignUp = values => ({
  type: AUTH_SIGN_UP,
  payload: values,
});

/**
 * Auth log out action
 *
 * @param {string} token
 * @returns {object}
 */
export const authLogOut = token => ({
  type: AUTH_LOG_OUT,
  payload: token,
});

/**
 * Auth set country action
 *
 * @param {string} countryCode
 * @returns {object}
 */
export const authSetCountryAction = countryCode => ({
  type: AUTH_COUNTRY_SET,
  payload: countryCode,
});

/**
 * Set remembered token
 *
 * @param {string} token
 */
export const setRememberedToken = token => localStorage.setItem(AUTH_TOKEN, token);

/**
 * Set remembered user
 *
 * @param {object} user
 */
export const setRememberedUser = user => localStorage.setItem(AUTH_USER, JSON.stringify(user));

/**
 * Set remembered country
 *
 * @param {string} countryCode
 */
export const setRememberedCountry = countryCode => localStorage.setItem(AUTH_COUNTRY_CODE, countryCode);

/**
 * Delete remembered token
 */
export const deleteRememberedToken = () => localStorage.removeItem(AUTH_TOKEN);

/**
 * Delete remembered user
 */
export const deleteRememberedUser = () => localStorage.removeItem(AUTH_USER);

/**
 * Delete remembered country
 */
export const deleteRememberedCountry = () => localStorage.removeItem(AUTH_COUNTRY_CODE);

/**
 * Auth set country action
 *
 * Sets the remembered country into local storage
 *
 * @param {string} countryCode
 * @returns {Function}
 */
export const authSetCountry = countryCode => dispatch => {
  setRememberedCountry(countryCode);
  dispatch(authSetCountryAction(countryCode));
};

/**
 * Get user action
 *
 * @param {Function} dispatch
 */
export const getUser = async dispatch => {
  dispatch(authRequest());

  try {
    const { data } = await client.getUser();
    dispatch(successAction(AUTH_REQUEST, data));
  } catch (e) {
    dispatch(errorFromResponse(AUTH_REQUEST, e, 'There was a problem fetching the current user.'));
  }
};

/**
 * Sign up action
 *
 * @param {object} values
 * @returns {Function}
 */
export const signUp = values => async dispatch => {
  dispatch(authSignUp(values));

  try {
    const { data } = await client.signUp({ data: { attributes: values } });
    dispatch(successAction(AUTH_SIGN_UP, data));
  } catch (e) {
    dispatch(errorFromResponse(AUTH_SIGN_UP, e));
  }
};

/**
 * Clear user action
 *
 * Removes user form local storage
 *
 * @param {Function} dispatch
 */
const clearUser = dispatch => {
  deleteRememberedToken();
  deleteRememberedUser();
  deleteRememberedCountry();

  dispatch(authReset());
  dispatch(successAction(AUTH_LOG_OUT, {}));
  dispatch(storeReset());
};

/**
 * Log out action
 *
 * Trrigger logout request
 *
 * @param {string} token
 * @returns {Function}
 */
export const logOut = token => async dispatch => {
  dispatch(authLogOut(token));

  clearUser(dispatch);

  try {
    await client.logOut();
  } catch (e) {
    return true;
  }
};

/**
 * Persist user action
 *
 * Sets user data to local storage
 *
 * @param {Function} dispatch
 * @param {object} data
 */
const persistUser = (dispatch, data) => {
  const userUuid = prop('uuid', data);

  // If a user is 'pending' then throw a forbidden error and return
  if (propEq('status', UserStatusTypes.PENDING, data)) {
    return dispatch(errorAction(AUTH_REQUEST, { status: FORBIDDEN, unverified: true }));
  }

  setRememberedToken(userUuid);
  setRememberedUser(data);
  setRememberedCountry(propOr('GB', 'countryCode', data));

  dispatch(setToken(userUuid));
};

/**
 * Log in action
 *
 * @param {object} values
 * @returns {Function}
 */
export const logIn = values => async dispatch => {
  dispatch(authRequest(omit(['password'], values)));

  try {
    const {
      data: { data },
    } = await client.logIn({ data: { attributes: values } });

    persistUser(dispatch, data);
    dispatch(successAction(AUTH_REQUEST, { user: { ...data } }));
  } catch (e) {
    dispatch(errorFromResponse(AUTH_REQUEST, e));
  }
};

/**
 * Reset password action
 *
 * @param {object} values
 * @returns {Function}
 */
export const resetPassword = values => async dispatch => {
  dispatch(authPasswordReset(values));

  try {
    await client.resetPassword({ data: { attributes: values } });
    dispatch(successAction(AUTH_PASSWORD_RESET, values));
  } catch (e) {
    dispatch(errorFromResponse(AUTH_SET_PASSWORD, e));
    throw e;
  }
};

/**
 * Set password action
 *
 * @param {object} values
 * @returns {Function}
 */
export const setPassword = values => async dispatch => {
  dispatch(authSetPasswordReset(omit(['values'], values)));

  try {
    await client.setPassword({ data: { attributes: values } });
    dispatch(successAction(AUTH_SET_PASSWORD, values));
  } catch (e) {
    dispatch(errorFromResponse(AUTH_SET_PASSWORD, e));
  }
};

/**
 * Auth check action
 *
 * @param {object} params
 * @returns {Function}
 */
export const authCheck = params => async (dispatch, getState) => {
  dispatch(genericAction(AUTH_CHECK));

  const role = getCurrentUserType(getState());

  // Associations for auth call are different per role
  // TAs : assignedSalesRepresentatives
  // Everyone else : assignedTravelAgents
  const authParams = mergeDeepLeft(params, {
    associations: equals(AuthTypes.TA, role) ? 'assignedSalesRepresentatives' : 'assignedTravelAgents',
  });

  try {
    const {
      data: { data },
    } = await userClient.me(authParams);
    dispatch(successAction(AUTH_CHECK, { user: { ...data } }));
    persistUser(dispatch, data);
  } catch (e) {
    clearUser(dispatch);

    throw e;
  }
};

/**
 * Update password action
 *
 * @param {object} body
 * @returns {Function}
 */
export const updatePassword = body => async dispatch => {
  dispatch(genericAction(AUTH_PASSWORD_UPDATE, body));

  try {
    await client.updatePassword({ data: { attributes: body } });
    dispatch(successAction(AUTH_SET_PASSWORD, body));
    dispatch(enqueueNotification({ message: 'Password has been updated.', options: { variant: 'success' } }));
  } catch (e) {
    dispatch(errorFromResponse(AUTH_PASSWORD_UPDATE, e));
  }
};
