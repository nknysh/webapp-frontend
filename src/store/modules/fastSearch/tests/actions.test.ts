import * as Actions from '../actions';
import { Filters, StarRating } from 'services/BackendApi';
import { sampleQuery } from 'services/BackendApi/tests/fixtures/sampleQuery';

describe('fastSearchActions Snapshot tests', () => {
  it('Matches snapshots', () => {
    // All actions return object literals, so snapshot testing just
    // gives us coverage.
    expect(Actions.initializeQueryAction('test')).toMatchSnapshot();
    expect(Actions.destinationChangeAction('hello')).toMatchSnapshot();
    expect(Actions.toggleFilterAction(Filters.BEST_FOR_FAMILIES)).toMatchSnapshot();
    expect(Actions.setFiltersAction([Filters.BEST_FOR_FAMILIES], true)).toMatchSnapshot();
    expect(Actions.setAllFiltersAction(true)).toMatchSnapshot();
    expect(Actions.toggleRepeatGuestAction()).toMatchSnapshot();
    expect(Actions.toggleStarRatingAction(StarRating.FiveStar)).toMatchSnapshot();
    expect(Actions.minPriceChangeAction(10)).toMatchSnapshot();
    expect(Actions.maxPriceChangeAction(100)).toMatchSnapshot();
    expect(Actions.incrementRoomAction(1)).toMatchSnapshot();
    expect(Actions.incrementAdultAction(1, 1)).toMatchSnapshot();
    expect(Actions.incrementChildAction(1, 1)).toMatchSnapshot();
    expect(Actions.setAgeAction(0, 0, '10')).toMatchSnapshot();
    expect(Actions.offersSearchRequestAction(sampleQuery)).toMatchSnapshot();
    expect(Actions.offersSearchSuccessAction('test' as any)).toMatchSnapshot();
    expect(Actions.offersSearchFailureAction('test' as any)).toMatchSnapshot();
  });
});

/*



  | ToggleStarRatingAction
  | ToggleOccasionAction
  | ToggleShowRegionsAction
  | ToggleRegionAction
  | ToggleHighlightsAction
  | ToggleLodgingControlsAction
  | SetLodgingControlsVisibilityAction
  | SelectMealPlanAction
  | SetFiltersAction
  | SetAllFiltersAction
  | MinPriceChangeAction
  | MaxPriceChangeAction
  | IncrementRoomAction
  | IncrementAdultAction
  | IncrementChildAction
  | SetAgeAction
  | SearchRequestAction
  | OffersSearchSuccessAction
  | OffersSearchFailureAction
  | OptionsRequestAction
  | OptionsSuccessAction
  | OptionsFailureAction
  | SetActiveLodgingIndexAction
  | IncrementActiveLodgingIndexAction
  | NamesSearchSuccessAction
  | NamesSearchFailureAction
  | SetNamesSearchResultsVisibilityAction
  | PopulateQueryAction
  | ClearExtendedQueryAction
  | DateRangeChangeAction
  | UpdateQueryStringAction
  | ResetSearchQueryAction;
*/
