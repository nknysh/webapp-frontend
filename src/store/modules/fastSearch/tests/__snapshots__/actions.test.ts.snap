// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fastSearchActions Snapshot tests Matches snapshots 1`] = `
Object {
  "type": "fastSearch/DESTINATION_CHANGE",
  "value": "hello",
}
`;

exports[`fastSearchActions Snapshot tests Matches snapshots 2`] = `
Object {
  "filter": "Best for Families",
  "type": "fastSearch/TOGGLE_FILTER",
}
`;

exports[`fastSearchActions Snapshot tests Matches snapshots 3`] = `
Object {
  "filters": Array [
    "Best for Families",
  ],
  "type": "fastSearch/SET_FILTER",
  "value": true,
}
`;

exports[`fastSearchActions Snapshot tests Matches snapshots 4`] = `
Object {
  "type": "fastSearch/SET_ALL_FILTERS",
  "value": true,
}
`;

exports[`fastSearchActions Snapshot tests Matches snapshots 5`] = `
Object {
  "type": "fastSearch/TOGGLE_REPEAT_GUEST",
}
`;

exports[`fastSearchActions Snapshot tests Matches snapshots 6`] = `
Object {
  "date": "2020-01-01",
  "type": "fastSearch/DATE_RANGE_START_CHANGE",
}
`;

exports[`fastSearchActions Snapshot tests Matches snapshots 7`] = `
Object {
  "date": "2020-01-07",
  "type": "fastSearch/DATE_RANGE_START_CHANGE",
}
`;

exports[`fastSearchActions Snapshot tests Matches snapshots 8`] = `
Object {
  "type": "fastSearch/MIN_PRICE_CHANGE",
  "value": 10,
}
`;

exports[`fastSearchActions Snapshot tests Matches snapshots 9`] = `
Object {
  "type": "fastSearch/MAX_PRICE_CHANGE",
  "value": 100,
}
`;

exports[`fastSearchActions Snapshot tests Matches snapshots 10`] = `
Object {
  "lodgingIndex": 1,
  "type": "fastSearch/INCREMENT_ROOM",
}
`;

exports[`fastSearchActions Snapshot tests Matches snapshots 11`] = `
Object {
  "lodgingIndex": 1,
  "type": "fastSearch/DECREMENT_ROOM",
}
`;

exports[`fastSearchActions Snapshot tests Matches snapshots 12`] = `
Object {
  "lodgingIndex": 1,
  "type": "fastSearch/INCREMENT_ADULT",
}
`;

exports[`fastSearchActions Snapshot tests Matches snapshots 13`] = `
Object {
  "lodgingIndex": 1,
  "type": "fastSearch/DECREMENT_ADULT",
}
`;

exports[`fastSearchActions Snapshot tests Matches snapshots 14`] = `
Object {
  "lodgingIndex": 1,
  "type": "fastSearch/INCREMENT_CHILD",
}
`;

exports[`fastSearchActions Snapshot tests Matches snapshots 15`] = `
Object {
  "lodgingIndex": 1,
  "type": "fastSearch/DECREMENT_CHILD",
}
`;

exports[`fastSearchActions Snapshot tests Matches snapshots 16`] = `
Object {
  "childIndex": 0,
  "lodgingIndex": 0,
  "type": "fastSearch/AGE_CHANGE",
  "value": "10",
}
`;

exports[`fastSearchActions Snapshot tests Matches snapshots 17`] = `
Object {
  "query": Object {
    "endDate": "2020-01-07",
    "filters": Array [
      "Seaplane transfer",
    ],
    "lodgings": Array [
      Object {
        "agesOfAllChildren": Array [],
        "honeymoon": true,
        "numberOfAdults": 1,
        "repeatCustomer": false,
      },
    ],
    "mealPlanCategories": Array [
      "BB",
    ],
    "name": "Amilla Fushi",
    "priceRange": Object {
      "max": 100000,
      "min": 1,
    },
    "regions": Array [],
    "starRatings": Array [
      "5+",
    ],
    "startDate": "2020-01-01",
  },
  "type": "fastSearch/OFFERS_SEARCH_REQUEST",
}
`;

exports[`fastSearchActions Snapshot tests Matches snapshots 18`] = `
Object {
  "successResponse": "test",
  "type": "fastSearch/OFFERS_SEARCH_SUCCESS",
}
`;

exports[`fastSearchActions Snapshot tests Matches snapshots 19`] = `
Object {
  "errorResponse": "test",
  "type": "fastSearch/OFFERS_SEARCH_FAILURE",
}
`;
